Generate heatmap transforms

eclipse0922 opened the pull request:
Fixes #3328 .

### Description

A few sentences describing the changes proposed in this pull request.

This pull request introduces `GenerateHeatmap` and `GenerateHeatmapd` transforms for creating Gaussian heatmaps from landmark coordinates.
The input points are currently expected in ZYX order, but this can be changed to support XYZ if preferred. 
The transforms ~support both batched (B, N, D) and~ only non-batched (N, D) inputs. 

Example notebooks are included for demonstration and will be removed before the PR is merged.


### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

Commit:
Adds a transform to generate heatmap from landmarks

Adds a `GenerateHeatmap` transform to create gaussian response maps from landmark coordinates.
This transform is implemented for both array and dictionary-based workflows.
It enables the generation of heatmaps from landmark data, facilitating tasks
like landmark localization and visualization.
The transform supports 2D and 3D coordinates and offers options for controlling
the gaussian standard deviation, spatial shape, truncation, normalization, and data type.

Commit:
Adds heatmap generation demo and tests

Introduces a new interactive notebook demonstrating landmark to heatmap conversion using MONAI transforms.

This includes:
- A notebook with array and dictionary transform modes.
- A test suite for the `GenerateHeatmap` transform.

This enhancement enables users to visualize and interact with heatmap generation, facilitating a better understanding and application of the MONAI transforms.

Commit:
Enables batched input for heatmap generation

Extends the `GenerateHeatmap` transform to support batched inputs,
allowing for more efficient processing of multiple landmark sets.

This change modifies the transform to handle inputs with a batch dimension (B, N, spatial_dims) in addition to single-point inputs (N, spatial_dims).
It also includes a demonstration of 3D heatmap generation using PyVista for visualization.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Refactors GenerateHeatmap transforms for clarity

Streamlines the GenerateHeatmap and GenerateHeatmapd transforms for better usability and code clarity.

Specifically:
- Improves the input landmark array validation to provide a more descriptive error message.
- Removes example notebooks.

DCO Remediation Commit for sewon.jeon <sewon.jeon@connecteve.com>

I, sewon.jeon <sewon.jeon@connecteve.com>, hereby add my Signed-off-by to this commit: 8ef905b36b26f2d84c386ab5359d76a61ca643ac
I, sewon.jeon <sewon.jeon@connecteve.com>, hereby add my Signed-off-by to this commit: 226bf906c9b62e3ca196ebd29dd81b0eede48b4a
I, sewon.jeon <sewon.jeon@connecteve.com>, hereby add my Signed-off-by to this commit: 3097baf5204b03872a87f8baa2e6d34411f9f42f
I, sewon.jeon <sewon.jeon@connecteve.com>, hereby add my Signed-off-by to this commit: 0072cb081dbe805cd4ed569c722f76f9721ab239

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
rename parameter

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
fix formatting

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
fix flake8

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
fix meta tensor problem

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
better unit tests

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

eclipse0922 commented:
[2d_mdtest.ipynb](https://github.com/user-attachments/files/22452287/2d_mdtest.ipynb)
[3d_heatmap_pyvista.ipynb](https://github.com/user-attachments/files/22452288/3d_heatmap_pyvista.ipynb)

Commit:
fix test error

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
address the code review.

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
fixes

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
Improves GenerateHeatmap transform and documentation

Enhances the GenerateHeatmap transform with better normalization,
spatial metadata handling, and comprehensive documentation.

The changes ensure correct heatmap normalization, and improve
handling of spatial metadata inheritance from reference images.
Also improves input validation and fixes shape inconsistencies.

Adds new test cases to cover edge cases and improve code reliability.
Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
Fixes heatmap normalization and shape checking

Fixes an issue where heatmap normalization was using the entire heatmap instead of the local region.

Adds a check to ensure that the provided static shape matches the number of spatial dimensions.

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
Adds `GenerateHeatmap` transform

Adds a `GenerateHeatmap` transform to generate heatmaps from point data.

This transform creates heatmaps from point data, validating that the dtype
is a floating-point type.

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
fix nitpick comments

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

eclipse0922 commented:
can anybod review this please?

eclipse0922 commented:
<img width="976" height="506" alt="image" src="https://github.com/user-attachments/assets/445b228f-457c-462c-8959-73ba2e10efba" />


[scene-export.html](https://github.com/user-attachments/files/22998595/scene-export.html)


2D and 3D example

eclipse0922 commented:
Could you please take a look at this PR when you have a moment? I believe the GenerateHeatmap would be a valuable addition. Thanks.

@KumoLiu @ericspod @Nic-Ma

ericspod commented on monai/transforms/post/array.py:
```diff
@@ -742,6 +750,194 @@ def __call__(self, img: Sequence[NdarrayOrTensor] | NdarrayOrTensor) -> NdarrayO
+        - Target spatial_shape is (Y, X) for 2D and (Z, Y, X) for 3D.
+        - Output layout uses channel-first convention with one channel per landmark:
+            - Non-batched points (N, D): (N, Y, X) for 2D or (N, Z, Y, X) for 3D
+            - Batched points (B, N, D): (B, N, Y, X) for 2D or (B, N, Z, Y, X) for 3D
```
Generally we don't expect batch dimensions in inputs to transforms. If the batch dimension is present is this assumed to be 0?

ericspod commented on monai/transforms/post/array.py:
```diff
@@ -742,6 +750,194 @@ def __call__(self, img: Sequence[NdarrayOrTensor] | NdarrayOrTensor) -> NdarrayO
+        - Output layout uses channel-first convention with one channel per landmark:
+            - Non-batched points (N, D): (N, Y, X) for 2D or (N, Z, Y, X) for 3D
+            - Batched points (B, N, D): (B, N, Y, X) for 2D or (B, N, Z, Y, X) for 3D
+        - Each channel corresponds to one landmark.
```
```suggestion
        - Each channel index corresponds to one landmark.
```

ericspod commented on monai/transforms/post/array.py:
```diff
@@ -742,6 +750,194 @@ def __call__(self, img: Sequence[NdarrayOrTensor] | NdarrayOrTensor) -> NdarrayO
+    Args:
+        sigma: gaussian standard deviation. A single value is broadcast across all spatial dimensions.
+        spatial_shape: optional fallback spatial shape. If ``None`` it must be provided when calling the transform.
+            A single int value will be broadcast to all spatial dimensions.
```
This doesn't seem to be allowed by the type of the argument nor make sense if it's ambiguous whether your input has a batch dimension or not.

ericspod commented on monai/transforms/post/array.py:
```diff
@@ -742,6 +750,194 @@ def __call__(self, img: Sequence[NdarrayOrTensor] | NdarrayOrTensor) -> NdarrayO
+            raise ValueError(f"dtype must be a floating-point type, got {self.torch_dtype}")
+        self.spatial_shape = None if spatial_shape is None else tuple(int(s) for s in spatial_shape)
+
+    def __call__(self, points: NdarrayOrTensor, spatial_shape: Sequence[int] | None = None) -> NdarrayOrTensor:
```
I realise a lot of computation is being done here specific to how you want to handle landmark heatmaps on a per landmark basis, but much of the code for doing the gaussian computation I think could be replaced with usages of [GaussianFilter](https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/layers/simplelayers.py#L542). The arguments for this class can be tweaked to get the same numeric result I think, or at least close enough for the purpose of heatmaps.

ericspod commented on monai/transforms/post/dictionary.py:
```diff
@@ -508,6 +513,175 @@ def __init__(self, keys: KeysCollection, output_key: str | None = None, num_clas
+    Converts landmark coordinates into gaussian heatmaps and optionally copies metadata from a reference image.
+
+    Args:
+        keys: keys of the corresponding items in the dictionary.
```
I think it needs to be clearer here that the keys are naming tensors of landmark points. An example in this docstring of using the class would help a lot.

ericspod commented on monai/transforms/post/dictionary.py:
```diff
@@ -508,6 +513,175 @@ def __init__(self, keys: KeysCollection, output_key: str | None = None, num_clas
+            # The GenerateHeatmap transform will handle type conversion based on input points
+            heatmap = self.generator(points, spatial_shape=shape)
+            # If there's a reference image and we need to match its type/device
+            reference = d.get(ref_key) if ref_key is not None and ref_key in d else None
```
```suggestion
            reference = d.get(ref_key, None)
```

I think this is equivalent.

ericspod commented on monai/transforms/post/dictionary.py:
```diff
@@ -508,6 +513,175 @@ def __init__(self, keys: KeysCollection, output_key: str | None = None, num_clas
+    _ERR_REF_KEYS_LEN = "ref_image_keys length must match keys length when provided."
+    _ERR_SHAPE_LEN = "spatial_shape length must match keys length when providing per-key shapes."
+    _ERR_NO_SHAPE = "Unable to determine spatial shape for GenerateHeatmapd. Provide spatial_shape or ref_image_keys."
+    _ERR_INVALID_POINTS = "landmark arrays must be 2D or 3D with shape (N, D) or (B, N, D)."
```
```suggestion
    _ERR_INVALID_POINTS = "Landmark arrays must be 2D or 3D with shape (N, D) or (B, N, D)."
```

ericspod commented on monai/transforms/post/array.py:
```diff
@@ -742,6 +750,194 @@ def __call__(self, img: Sequence[NdarrayOrTensor] | NdarrayOrTensor) -> NdarrayO
+        self.numpy_dtype = get_equivalent_dtype(dtype, np.ndarray)
+        # Validate that dtype is floating-point for meaningful Gaussian values
+        if not self.torch_dtype.is_floating_point:
+            raise ValueError(f"dtype must be a floating-point type, got {self.torch_dtype}")
```
```suggestion
            raise ValueError(f"Argument `dtype` must be a floating-point type, got {self.torch_dtype}")
```
I'd suggest making it clear the word "dtype" refers to an argument, this applies elsewhere in error strings.

ericspod commented on tests/transforms/test_generate_heatmapd.py:
```diff
@@ -0,0 +1,267 @@
+        torch.float32,
+        True,  # uses reference image
+    ]
+)
```
```suggestion
TEST_CASES_WITH_REF = [
    [
        "dict_with_ref_3d_numpy",
        np.array([[2.5, 2.5, 3.0], [5.0, 5.0, 4.0]], dtype=np.float32),
        {"sigma": 2.0},
        (2, 8, 8, 8),
        torch.float32,
        True,  # uses reference image
    ],
    [
        "dict_with_ref_3d_torch",
        torch.tensor([[2.5, 2.5, 3.0], [5.0, 5.0, 4.0]], dtype=torch.float32),
        {"sigma": 2.0},
        (2, 8, 8, 8),
        torch.float32,
        True,  # uses reference image
    ]
]
```
I think it's fine to define lists directly like this.

ericspod commented on tests/transforms/test_generate_heatmap.py:
```diff
@@ -0,0 +1,312 @@
+            {"sigma": 1.5, "spatial_shape": (16, 16)},
+            (2, 16, 16),
+        ]
+    )
```
```suggestion
TEST_CASES_2D = [
    [
        f"2d_basic_type{idx}",
        p(np.array([[4.2, 7.8], [12.3, 3.6]], dtype=np.float32)),
        {"sigma": 1.5, "spatial_shape": (16, 16)},
        (2, 16, 16),
    ]
    for idx, p in enumerate(TEST_NDARRAYS)
]
```
If you wanted I think defining values like this is fine.

ericspod commented on tests/transforms/test_generate_heatmap.py:
```diff
@@ -0,0 +1,312 @@
+                dtype,
+                "cpu",
+            ]
+        )
```
```suggestion
test_device = "cuda:0" if torch.cuda.is_available() else "cpu"
test_dtypes = [torch.float32, torch.float64]
if torch.cuda.is_available():
    test_dtypes.append(torch.float16)
    
TEST_CASES_DEVICE_DTYPE = [
    [
        f"cuda_{str(dtype).replace('torch.', '')}",
        torch.tensor([[3.0, 4.0, 5.0]], dtype=torch.float32, device=test_device),
        {"sigma": 1.2, "spatial_shape": (10, 10, 10), "dtype": dtype},
        (1, 10, 10, 10),
        dtype,
        test_device,
    ]
    for dtype in test_dtypes
]
```

ericspod reviewed (changes requested):
Hi @eclipse0922 sorry for the delay in reviewing. I think these classes would be a good addition to MONAI, I had a few comments below but I hadn't gone through the code thoroughly quiet yet. I commented below that these classes could use [GaussianFilter](https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/layers/simplelayers.py#L542) instead of defining the gaussian computation over again, this would save reduce line count and so complexity, and should be adaptable to how you're computing the heatmaps. Please have a look at that class, or elsewhere in MONAI that gaussian kernels are used. Thanks!

eclipse0922 commented:
> Hi @eclipse0922 sorry for the delay in reviewing. I think these classes would be a good addition to MONAI, I had a few comments below but I hadn't gone through the code thoroughly quiet yet. I commented below that these classes could use [GaussianFilter](https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/layers/simplelayers.py#L542) instead of defining the gaussian computation over again, this would save reduce line count and so complexity, and should be adaptable to how you're computing the heatmaps. Please have a look at that class, or elsewhere in MONAI that gaussian kernels are used. Thanks!

@ericspod 

 Thank you for the  detailed review. I'll look into your comments and incorporate the changes, including the suggestion to use GaussianFilter

Commit:
Remove batch, use GaussianFilter

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
fix the comment by ai review

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
fix dtype

DCO Remediation Commit for sewon.jeon <sewon.jeon@connecteve.com>

I, sewon.jeon <sewon.jeon@connecteve.com>, hereby add my Signed-off-by to this commit: 8e1910b03c2a575b7d7ff323468d2f327f8f2025
I, sewon.jeon <sewon.jeon@connecteve.com>, hereby add my Signed-off-by to this commit: c0c0a84c3236203b132ea48c421d7f01f6bc8220
I, sewon.jeon <sewon.jeon@connecteve.com>, hereby add my Signed-off-by to this commit: 7c988921341790a45d49a09ef69d0e3f233e598f

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

eclipse0922 commented:
[2d_mdtest.ipynb](https://github.com/user-attachments/files/23337943/2d_mdtest.ipynb)
[3d_heatmap_pyvista.ipynb](https://github.com/user-attachments/files/23337944/3d_heatmap_pyvista.ipynb)
updated notebook

eclipse0922 committed:
Update monai/transforms/post/dictionary.py

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>
Signed-off-by: sewon jeon <irocks0922@gmail.com>

eclipse0922 commented:
Hi @ericspod, just a friendly reminder regarding this PR.

I have addressed your feedback from the last review, specifically refactoring the code to use `GaussianFilter` instead of the custom implementation. I also fixed the issues reported by the AI review bots.

Could you please take a look again when you have a chance? Thanks!

ericspod reviewed (approved):
Hi @eclipse0922 I think looks good now. We can merge this soon. Thanks!

ericspod committed:
Merge branch 'dev' into generate_heatmap_transforms

KumoLiu committed:
Merge branch 'dev' into generate_heatmap_transforms

KumoLiu commented:
/build
